File: src/main/java/eu/mshade/enderchest/marshal/world/ChunkBinaryTagMarshal.java
package eu.mshade.enderchest.marshal.world;

import eu.mshade.enderchest.world.DefaultChunk;
import eu.mshade.enderframe.world.Chunk;
import eu.mshade.enderframe.world.Section;
import eu.mshade.enderframe.world.World;
import eu.mshade.mwork.binarytag.BinaryTag;
import eu.mshade.mwork.binarytag.BinaryTagDriver;
import eu.mshade.mwork.binarytag.BinaryTagDynamicMarshal;
import eu.mshade.mwork.binarytag.BinaryTagType;
import eu.mshade.mwork.binarytag.entity.ByteArrayBinaryTag;
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag;
import eu.mshade.mwork.binarytag.entity.ListBinaryTag;
import eu.mshade.mwork.binarytag.entity.ZstdCompoundBinaryTag;
import eu.mshade.mwork.binarytag.poet.BinaryTagPoet;

import java.io.IOException;
import java.util.concurrent.ExecutionException;

public class ChunkBinaryTagMarshal implements BinaryTagDynamicMarshal {


    public CompoundBinaryTag serialize(BinaryTagDriver binaryTagDriver, Chunk chunk) {


        CompoundBinaryTag compoundBinaryTag = new ZstdCompoundBinaryTag();

        compoundBinaryTag.putInt("x", chunk.getX());
        compoundBinaryTag.putInt("z", chunk.getZ());
        compoundBinaryTag.putBinaryTag("biomes", new ByteArrayBinaryTag(chunk.getBiomes()));

        ListBinaryTag listBinaryTagSections = new ListBinaryTag(BinaryTagType.COMPOUND);
        ListBinaryTag listBinaryTagEntities = new ListBinaryTag(BinaryTagType.COMPOUND);

        SectionBinaryTagMarshal sectionBinaryTagMarshal = binaryTagDriver.getDynamicMarshal(SectionBinaryTagMarshal.class);

        for (Section section : chunk.getSections()) {
            if (section != null) listBinaryTagSections.add(sectionBinaryTagMarshal.serialize(section));
        }

        /*
        for(Entity entity : chunk.getEntities()) {
            listBinaryTagEntities.add(binaryTagDriver.marshal(entity));
        }

         */

        compoundBinaryTag.putBinaryTag("sections", listBinaryTagSections);
        compoundBinaryTag.putBinaryTag("entities", listBinaryTagEntities);
        return compoundBinaryTag;
    }

    public Chunk deserialize(BinaryTagDriver binaryTagDriver, BinaryTag<?> binaryTag, World world) {
        CompoundBinaryTag compoundBinaryTag = (CompoundBinaryTag) binaryTag;
        SectionBinaryTagMarshal sectionBinaryTagMarshal = binaryTagDriver.getDynamicMarshal(SectionBinaryTagMarshal.class);

        int x = compoundBinaryTag.getInt("x");

        int z = compoundBinaryTag.getInt("z");

        byte[] biome = compoundBinaryTag.getByteArray("biomes");
        ListBinaryTag sectionBinaryTags = (ListBinaryTag) compoundBinaryTag.getBinaryTag("sections");
        ListBinaryTag entityBinaryTags = (ListBinaryTag) compoundBinaryTag.getBinaryTag("entities");

        Chunk chunk = new DefaultChunk(x, z, world, biome);
        Section[] sections = chunk.getSections();

        sectionBinaryTags.forEach(sectionBinaryTag -> {
            Section section = sectionBinaryTagMarshal.deserialize(sectionBinaryTag, chunk);
            sections[section.getY()] = section;
        });

        /*
        entityBinaryTags.forEach(entityBinaryTag ->{
            CompoundBinaryTag compoundBinaryTagEntity = (CompoundBinaryTag)entityBinaryTag;
            EntityType entityType = EntityType.getEntityTypeByName(compoundBinaryTagEntity.getString("entityType"));
            chunk.addEntity(binaryTagDriver.unMarshal(entityBinaryTag, entityType.getClazz()));
        });

         */
        return chunk;
    }

    private String regionId(Chunk chunk){
        return regionId(chunk.getX(), chunk.getZ());
    }

    private String regionId(int chunkX, int chunkZ){
        return (chunkX >> 5) +","+(chunkZ >> 5);
    }

    private String chunkId(Chunk chunk){
        return chunk.getX()+","+chunk.getZ();
    }
    private String chunkId(int chunkX, int chunkZ){
        return chunkX+","+chunkZ;
    }


    public void write(BinaryTagDriver binaryTagDriver, BinaryTagPoet binaryTagPoet, Chunk chunk){
        try {
            binaryTagPoet.writeCompoundBinaryTag(chunkId(chunk), this.serialize(binaryTagDriver, chunk));
        } catch (ExecutionException | InterruptedException | IOException e) {
            e.printStackTrace();
        }
    }

    public Chunk read(BinaryTagDriver binaryTagDriver, BinaryTagPoet binaryTagPoet, World world, int chunkX, int chunkZ) throws IOException {
        String chunkId = chunkId(chunkX, chunkZ);
        CompoundBinaryTag compoundBinaryTag = binaryTagPoet.readCompoundBinaryTag(chunkId);
        return deserialize(binaryTagDriver, compoundBinaryTag, world);
    }


}

File: src/main/java/eu/mshade/enderchest/marshal/world/ChunkBinaryTagMarshal.kt
package eu.mshade.enderchest.marshal.world

import eu.mshade.enderchest.EnderChest
import eu.mshade.enderframe.metadata.MetadataKeyValueBufferRegistry
import eu.mshade.enderchest.world.DefaultChunk
import eu.mshade.enderframe.MinecraftServer
import eu.mshade.enderframe.inventory.Inventory
import eu.mshade.enderframe.world.Vector
import eu.mshade.enderframe.world.World
import eu.mshade.enderframe.world.block.BlockMetadataType
import eu.mshade.enderframe.world.chunk.Chunk
import eu.mshade.mwork.binarytag.*
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag
import eu.mshade.mwork.binarytag.entity.ListBinaryTag
import eu.mshade.mwork.binarytag.segment.SegmentBinaryTag
import java.io.IOException
import java.util.concurrent.ExecutionException

object ChunkBinaryTagMarshal {

    val tickableBlocks = MinecraftServer.getTickableBlocks()

    fun serialize(
        chunk: Chunk,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): CompoundBinaryTag {
        val compoundBinaryTag: CompoundBinaryTag = ZstdCompoundBinaryTag()
        compoundBinaryTag.putInt("x", chunk.x)
        compoundBinaryTag.putInt("z", chunk.z)
        compoundBinaryTag.putBinaryTag("biomes", ByteArrayBinaryTag(chunk.biomes))
        val listBinaryTagSections = ListBinaryTag(BinaryTagType.COMPOUND)
        val listBinaryTagEntities = ListBinaryTag(BinaryTagType.COMPOUND)

        for (i in chunk.sections.indices) {
            val section = chunk.sections[i]
            if (section != null) {
                listBinaryTagSections.add(SectionBinaryTagMarshal.serialize(section, metadataKeyValueBufferRegistry))
            }
        }

        /*
    for(Entity entity : chunk.getEntities()) {
        listBinaryTagEntities.add(binaryTagDriver.marshal(entity));
    }

     */compoundBinaryTag.putBinaryTag("sections", listBinaryTagSections)
        compoundBinaryTag.putBinaryTag("entities", listBinaryTagEntities)
        return compoundBinaryTag
    }

    fun deserialize(
        binaryTag: BinaryTag<*>,
        world: World,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): Chunk {
        val compoundBinaryTag = binaryTag as CompoundBinaryTag?

        val x = compoundBinaryTag!!.getInt("x")
        val z = compoundBinaryTag.getInt("z")
        val biome = compoundBinaryTag.getByteArray("biomes")
        val sectionBinaryTags = compoundBinaryTag.getBinaryTag("sections") as ListBinaryTag
        val entityBinaryTags = compoundBinaryTag.getBinaryTag("entities") as ListBinaryTag
        val chunk: Chunk = DefaultChunk(x, z, world)
        chunk.biomes = biome
        val sections = chunk.sections
        sectionBinaryTags.value.forEach { sectionBinaryTag ->
            val section =
                SectionBinaryTagMarshal.deserialize(sectionBinaryTag, chunk, metadataKeyValueBufferRegistry)
            sections[section.y] = section
        }

        for (section in sections) {
            if (section == null) continue
            for (i in 0 until section.blocks.size) {
                val block = section.getBlock(i)

                if (block.isTickable() && block.getPower() == 15) {
                    val vectorFromIndex = chunk.getVectorFromIndex(i)
                    val position = Vector(
                        vectorFromIndex.x + x * 16,
                        vectorFromIndex.y + section.y * 16,
                        vectorFromIndex.z + z * 16
                    )
                    tickableBlocks.join(world, position)
                }
            }
        }


        /*
    entityBinaryTags.forEach(entityBinaryTag ->{
        CompoundBinaryTag compoundBinaryTagEntity = (CompoundBinaryTag)entityBinaryTag;
        EntityType entityType = EntityType.getEntityTypeByName(compoundBinaryTagEntity.getString("entityType"));
        chunk.addEntity(binaryTagDriver.unMarshal(entityBinaryTag, entityType.getClazz()));
    });

     */return chunk
    }

    private fun regionId(chunk: Chunk): String {
        return regionId(chunk.x, chunk.z)
    }

    private fun regionId(chunkX: Int, chunkZ: Int): String {
        return (chunkX shr 5).toString() + "," + (chunkZ shr 5)
    }

    private fun chunkId(chunk: Chunk): String {
        return chunk.x.toString() + "," + chunk.z
    }

    private fun chunkId(chunkX: Int, chunkZ: Int): String {
        return "$chunkX,$chunkZ"
    }

    fun write(
        segment: SegmentBinaryTag,
        binaryTagDriver: BinaryTagDriver,
        chunk: Chunk,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ) {
        try {
            segment.writeCompound(
                chunkId(chunk),
                binaryTagDriver,
                this.serialize(chunk, metadataKeyValueBufferRegistry)
            )
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun read(
        segment: SegmentBinaryTag,
        binaryTagDriver: BinaryTagDriver,
        world: World,
        chunkX: Int,
        chunkZ: Int,
        metadataKeyValueBufferRegistry: MetadataKeyValueBufferRegistry
    ): Chunk {
        val chunkId = chunkId(chunkX, chunkZ)
        val compoundBinaryTag = segment.readCompound(chunkId, binaryTagDriver)
        return deserialize(compoundBinaryTag, world, metadataKeyValueBufferRegistry)
    }

}